proposals/simd/portable-simd.md

```diff
@@ -0,0 +1,1320 @@
+| `s16x8` | -2^15 | 2^15-1 |
+| `s32x4` | -2^31 | 2^31-1 |
+| `s64x2` | -2^63 | 2^63-1 |
+
+### Unsigned integer interpretations
+
+Each lane is interpreted as an unsigned integer.
+
+* `u8x16 : i8x16`: Lane values in the range 0 – 2^8-1.
+* `u16x8 : i16x8`: Lane values in the range 0 – 2^16-1.
+* `u32x4 : i32x4`: Lane values in the range 0 – 2^32-1.
+* `u64x2 : i64x2`: Lane values in the range 0 – 2^64-1.
+
+These interpretations get additional properties defining the range of values in
+a lane:
+
+|    S    | S.Min | S.Max  |
+|---------|------:|-------:|
+| `u8x16` |     0 | 2^8-1  |
+| `u16x8` |     0 | 2^16-1 |
+| `u32x4` |     0 | 2^32-1 |
+| `u64x2` |     0 | 2^64-1 |
+
+### Floating-point interpretations
+
+Each lane is interpreted as an IEEE floating-point number.
+
+* `f32x4 : v32x4`: Each lane is an `f32`.
+* `f64x2 : v64x2`: Each lane is an `f64`.
+
+Additional properties:
+
+|    S    | S.LaneType |
+|---------|------------|
+| `f32x4` | `f32`      |
+| `f64x2` | `f64`      |
+
+# Floating-point semantics
+
+The floating-point operations in this specification aim to be conforming to
+[IEEE 754-2008][ieee754] while being compatible with WebAssembly and JavaScript.
+Some things which are left unspecified by the IEEE standard are given stricter
+semantics by WebAssembly.
+
+## Rounding modes
+
+This specification does not yet provide a way of changing floating point
+rounding modes. All floating point operations use the default *roundTiesToEven*
+mode.
+
+## Default NaN value
+
+When a floating-point operation needs to return a NaN and none of its operands
+are NaN, it generates a default NaN value which is a quiet NaN with an all-zero
+payload field. The sign of the default NaN is not specified:
+
+```python
+def f32.default_nan():
+    if unspecified_choice():
+        bits = 0x7fc00000
+    else:
+        bits = 0xffc00000
+    return f32.from_bits(bits)
+
+def f64.default_nan():
+    if unspecified_choice():
+        bits = 0x7ff8000000000000
+    else:
+        bits = 0xfff8000000000000
+    return f64.from_bits(bits)
+```
+
+## Propagating NaN values
+
+When propagating a NaN value from an operand, all the bits of the NaN are
+preserved, except a signaling NaN is quieted by setting the most significand
+bit in the trailing significand field.
+
+```python
+def canonicalize_nan(x):
+    assert isnan(x)
+    t = type(x)
+    assert t == f32 or t == f64
+    bits = x.to_bits()
+    if t == f32:
+        bits |= (1 << 22)
+    else:
+        bits |= (1 << 51)
+    return t.from_bits(bits)
+```
+
+When two operands are NaN, one of them is propagated. Which one is not specified:
+
+```python
+def propagate_nan(x, y):
+    assert isinan(x) or isnan(y)
+    if not isnan(x):
+        return canonicalize_nan(y)
+    if not isnan(y)
+        return canonicalize_nan(x)
+    # Both x and y are NaNs: pick one to propagate.
+    if unspecified_choice():
+        return canonicalize_nan(x)
+    else:
+        return canonicalize_nan(y)
```

Comment:
That's what you get for specifying things twice. I think the next step for this proposal is to roll up this portable spec and the wasm specifics into a single wasm-only document. Then I can simply refer to existing semantics instead of trying to do a stand-alone spec.